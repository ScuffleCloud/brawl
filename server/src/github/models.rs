use chrono::{DateTime, Utc};
use octocrab::models::pulls::{MergeableState, ReviewState};
use octocrab::models::{InstallationId, IssueState, RepositoryId, RunId, UserId, WorkflowId};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct User {
    pub login: String,
    pub id: UserId,
}

impl Default for User {
    fn default() -> Self {
        Self {
            login: "user".to_string(),
            id: UserId(0),
        }
    }
}

impl From<octocrab::models::UserProfile> for User {
    fn from(value: octocrab::models::UserProfile) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

impl From<octocrab::models::Author> for User {
    fn from(value: octocrab::models::Author) -> Self {
        Self {
            login: value.login,
            id: value.id,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Label {
    pub name: String,
}

impl From<octocrab::models::Label> for Label {
    fn from(value: octocrab::models::Label) -> Self {
        Self { name: value.name }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct PrBranch {
    pub label: Option<String>,
    pub ref_field: String,
    pub sha: String,
    pub user: Option<User>,
    pub repo: Option<Repository>,
}

impl From<octocrab::models::pulls::Head> for PrBranch {
    fn from(value: octocrab::models::pulls::Head) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

impl From<octocrab::models::pulls::Base> for PrBranch {
    fn from(value: octocrab::models::pulls::Base) -> Self {
        Self {
            label: value.label,
            ref_field: value.ref_field,
            sha: value.sha,
            user: value.user.map(|u| u.into()),
            repo: value.repo.map(|r| r.into()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct PullRequest {
    pub number: u64,
    pub title: String,
    pub body: String,
    pub state: Option<IssueState>,
    pub mergeable_state: Option<MergeableState>,
    pub merge_commit_sha: Option<String>,
    pub assignees: Vec<User>,
    pub requested_reviewers: Vec<User>,
    pub user: Option<User>,
    pub labels: Vec<Label>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
    pub draft: Option<bool>,
    pub closed_at: Option<DateTime<Utc>>,
    pub merged_at: Option<DateTime<Utc>>,
    pub head: PrBranch,
    pub base: PrBranch,
}

impl From<octocrab::models::pulls::PullRequest> for PullRequest {
    fn from(value: octocrab::models::pulls::PullRequest) -> Self {
        Self {
            number: value.number,
            title: value.title.unwrap_or_default(),
            body: value.body.unwrap_or_default(),
            state: value.state,
            mergeable_state: value.mergeable_state,
            merge_commit_sha: value.merge_commit_sha,
            assignees: value.assignees.into_iter().flatten().map(|a| a.into()).collect(),
            requested_reviewers: value.requested_reviewers.into_iter().flatten().map(|a| a.into()).collect(),
            user: value.user.map(|u| (*u).into()),
            labels: value.labels.into_iter().flatten().map(|l| l.into()).collect(),
            created_at: value.created_at,
            updated_at: value.updated_at,
            draft: value.draft,
            closed_at: value.closed_at,
            merged_at: value.merged_at,
            head: (*value.head).into(),
            base: (*value.base).into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Default)]
pub struct Review {
    pub user: Option<User>,
    pub state: Option<ReviewState>,
}

impl From<octocrab::models::pulls::Review> for Review {
    fn from(value: octocrab::models::pulls::Review) -> Self {
        Self {
            user: value.user.map(|u| u.into()),
            state: value.state,
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct Commit {
    pub sha: String,
    pub tree: Option<String>,
}

impl From<octocrab::models::commits::Commit> for Commit {
    fn from(value: octocrab::models::commits::Commit) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.commit.tree.sha),
        }
    }
}

impl From<octocrab::models::commits::GitCommitObject> for Commit {
    fn from(value: octocrab::models::commits::GitCommitObject) -> Self {
        Self {
            sha: value.sha,
            tree: Some(value.tree.sha),
        }
    }
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunStatus {
    #[default]
    Queued,
    InProgress,
    Completed,
    Waiting,
    Requested,
    Pending,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
#[serde(rename_all = "snake_case")]
pub enum CheckRunConclusion {
    Success,
    Failure,
    #[default]
    Neutral,
    Cancelled,
    Skipped,
    TimedOut,
    ActionRequired,
}

#[derive(Debug, Clone, serde::Deserialize, PartialEq, Eq, Default)]
pub struct CheckRunEvent {
    pub id: i64,
    pub name: String,
    pub head_sha: String,
    pub html_url: Option<String>,
    pub details_url: Option<String>,
    pub url: String,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub status: CheckRunStatus,
    pub conclusion: Option<CheckRunConclusion>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Repository {
    pub id: RepositoryId,
    pub name: String,
    pub owner: User,
    pub default_branch: Option<String>,
}

impl From<octocrab::models::Repository> for Repository {
    fn from(value: octocrab::models::Repository) -> Self {
        Self {
            id: value.id,
            name: value.name,
            owner: value.owner.expect("repository has no owner").into(),
            default_branch: value.default_branch,
        }
    }
}

impl Default for Repository {
    fn default() -> Self {
        Self {
            id: RepositoryId(0),
            name: "repo".to_string(),
            owner: User::default(),
            default_branch: Some("main".to_string()),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Installation {
    pub id: InstallationId,
    pub account: User,
}

impl Default for Installation {
    fn default() -> Self {
        Self {
            id: InstallationId(0),
            account: User::default(),
        }
    }
}

impl From<octocrab::models::Installation> for Installation {
    fn from(value: octocrab::models::Installation) -> Self {
        Self {
            id: value.id,
            account: value.account.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct WorkflowRun {
    pub id: RunId,
    pub workflow_id: WorkflowId,
    pub name: String,
    pub head_sha: String,
    pub head_branch: String,
    pub conclusion: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub status: String,
}

impl From<octocrab::models::workflows::Run> for WorkflowRun {
    fn from(value: octocrab::models::workflows::Run) -> Self {
        Self {
            id: value.id,
            workflow_id: value.workflow_id,
            name: value.name,
            head_sha: value.head_sha,
            head_branch: value.head_branch,
            conclusion: value.conclusion,
            created_at: value.created_at,
            updated_at: value.updated_at,
            status: value.status,
        }
    }
}

impl Default for WorkflowRun {
    fn default() -> Self {
        Self {
            id: RunId(0),
            workflow_id: WorkflowId(0),
            name: "workflow".to_string(),
            head_sha: "sha".to_string(),
            head_branch: "branch".to_string(),
            conclusion: None,
            created_at: Utc::now(),
            updated_at: Utc::now(),
            status: "status".to_string(),
        }
    }
}
