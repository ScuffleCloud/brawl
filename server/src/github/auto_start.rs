use std::collections::HashMap;

use anyhow::Context;
use diesel_async::pooled_connection::bb8;
use diesel_async::AsyncPgConnection;
use octocrab::models::RepositoryId;
use scuffle_context::ContextFutExt;

use super::GitHubService;
use crate::schema::ci::CiRun;
use crate::schema::pr::Pr;

pub struct AutoStartSvc;

pub trait AutoStartConfig: Send + Sync + 'static {
	fn github_service(&self) -> &GitHubService;

	fn database_pool(&self) -> &bb8::Pool<AsyncPgConnection>;

	fn interval(&self) -> std::time::Duration;
}

impl<C: AutoStartConfig> scuffle_bootstrap::Service<C> for AutoStartSvc {
	async fn enabled(&self, _: &std::sync::Arc<C>) -> anyhow::Result<bool> {
		Ok(true)
	}

	async fn run(self, global: std::sync::Arc<C>, ctx: scuffle_context::Context) -> anyhow::Result<()> {
		tracing::info!("starting auto start service");

		while tokio::time::sleep(global.interval()).with_context(&ctx).await.is_some() {
			let mut conn = global.database_pool().get().await.context("get database connection")?;
			let github_service = global.github_service();

			let runs = CiRun::find_pending_runs(&mut conn).await?;
			let mut run_map = HashMap::<_, &CiRun>::new();
			for run in &runs {
				run_map
					.entry((run.github_repo_id, run.ci_branch.as_str()))
					.and_modify(|current| {
						let higher_priority = run.started_at.is_some()
							|| (run.priority, run.created_at) > (current.priority, current.created_at);

						if higher_priority && current.started_at.is_none() {
							*current = run;
						}
					})
					.or_insert(run);
			}

			for (_, run) in run_map {
				if let Err(e) = handle_run(run, github_service, &mut conn).await {
					tracing::error!(
						"error handling run (repo id: {}, pr number: {}, run id: {}): {}",
						run.github_repo_id,
						run.github_pr_number,
						run.id,
						e
					);
				}
			}
		}

		Ok(())
	}
}

async fn handle_run(run: &CiRun, github_service: &GitHubService, conn: &mut AsyncPgConnection) -> anyhow::Result<()> {
	let repo_id = RepositoryId(run.github_repo_id as u64);

	let Some(installation_client) = github_service.get_client_by_repo(repo_id) else {
		anyhow::bail!("no installation client found for repo {}", run.github_repo_id);
	};

	let config = installation_client
		.get_repo_config(repo_id)
		.await
		.context("get repo config")?;

	let Some(pr) = Pr::fetch(conn, repo_id, run.github_pr_number as i64)
		.await
		.context("fetch pr")?
	else {
		anyhow::bail!("no pr found for run");
	};

	if run.started_at.is_none() {
		run.start(conn, &installation_client, &config, &pr)
			.await
			.context("start run")?;
	} else {
		run.refresh(conn, &installation_client, &config, &pr)
			.await
			.context("refresh run")?;
	}

	Ok(())
}
