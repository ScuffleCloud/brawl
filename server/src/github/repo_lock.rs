use std::collections::HashMap;
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use octocrab::models::RepositoryId;
use parking_lot::Mutex;
use tokio::sync::OwnedSemaphorePermit;

#[derive(Debug, Clone)]
pub struct RepoLock {
    active_repos: Arc<Mutex<HashMap<RepositoryId, Lock>>>,
}

#[derive(Debug, Clone)]
struct Lock {
    count: Arc<AtomicUsize>,
    semaphore: Arc<tokio::sync::Semaphore>,
}

pub struct LockGuard {
    _inner: LockGuardInner,
    _permit: OwnedSemaphorePermit,
}

struct LockGuardInner {
    repo_id: RepositoryId,
    count: Arc<AtomicUsize>,
    repo_lock: RepoLock,
}

impl LockGuardInner {
    fn new(repo_id: RepositoryId, count: Arc<AtomicUsize>, repo_lock: RepoLock) -> Self {
        count.fetch_add(1, Ordering::Relaxed);
        Self {
            repo_id,
            count,
            repo_lock,
        }
    }

    fn into_guard(self, permit: OwnedSemaphorePermit) -> LockGuard {
        LockGuard {
            _inner: self,
            _permit: permit,
        }
    }
}

impl Drop for LockGuardInner {
    fn drop(&mut self) {
        if self.count.fetch_sub(1, Ordering::Relaxed) == 1 {
            let mut lock = self.repo_lock.active_repos.lock();
            // The reason we do an extra load here is that the mutex is required to do a
            // increment, therefore its possible that the count might have been
            // incremented by another thread after we decremented it to 0, but
            // before we could aquire the lock. So the additional check is to ensure we
            // the count is 0 before deleting the entry.
            if self.count.load(Ordering::Relaxed) == 0 {
                lock.remove(&self.repo_id);
            }
        }
    }
}

impl Lock {
    fn new() -> Self {
        Self {
            count: Arc::new(AtomicUsize::new(0)),
            semaphore: Arc::new(tokio::sync::Semaphore::new(1)),
        }
    }
}

impl Default for RepoLock {
    fn default() -> Self {
        Self::new()
    }
}

impl RepoLock {
    pub fn new() -> Self {
        Self {
            active_repos: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    pub async fn lock(&self, repo_id: RepositoryId) -> LockGuard {
        let (inner, semaphore) = {
            let mut lock = self.active_repos.lock();
            let lock = lock.entry(repo_id).or_insert_with(Lock::new);
            let inner = LockGuardInner::new(repo_id, lock.count.clone(), self.clone());
            (inner, lock.semaphore.clone())
        };

        inner.into_guard(semaphore.acquire_owned().await.expect("semaphore has no permits"))
    }
}
