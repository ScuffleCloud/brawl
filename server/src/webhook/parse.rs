use axum::body::Body;
use axum::http::{HeaderMap, HeaderValue, Request, StatusCode};
use hmac::{Hmac, Mac};
use octocrab::models::webhook_events::{EventInstallation, WebhookEventPayload, WebhookEventType};
use octocrab::models::InstallationId;
use sha2::Sha256;

use crate::github::models::{Installation, Repository, User};

fn verify_gh_signature(headers: &HeaderMap<HeaderValue>, body: &[u8], secret: &str) -> bool {
    let Some(signature) = headers.get("x-hub-signature-256").map(|v| v.as_bytes()) else {
        return false;
    };
    let Some(signature) = signature.get(b"sha256=".len()..).and_then(|v| hex::decode(v).ok()) else {
        return false;
    };

    let mut mac = Hmac::<Sha256>::new_from_slice(secret.as_bytes()).expect("Cannot create HMAC key");
    mac.update(body);
    mac.verify_slice(&signature).is_ok()
}

#[derive(Debug, Clone)]
pub struct WebhookEvent {
    pub sender: Option<User>,
    pub repository: Option<Repository>,
    pub installtion_id: Option<InstallationId>,
    pub installation: Option<Installation>,
    pub kind: WebhookEventType,
    pub specific: WebhookEventPayload,
}

fn parse_event(header: String, body: &[u8]) -> Result<WebhookEvent, (StatusCode, String)> {
    // NOTE: this is inefficient code to simply reuse the code from "derived"
    // serde::Deserialize instead of writing specific deserialization code for the
    // enum.
    let kind: WebhookEventType =
        serde_json::from_value(serde_json::Value::String(header)).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Intermediate structure allows to separate the common fields from
    // the event specific one.
    #[derive(serde::Deserialize)]
    struct Intermediate {
        sender: Option<octocrab::models::Author>,
        repository: Option<octocrab::models::Repository>,
        #[allow(unused)]
        organization: Option<octocrab::models::orgs::Organization>,
        installation: Option<octocrab::models::webhook_events::EventInstallation>,
        #[serde(flatten)]
        specific: serde_json::Value,
    }

    let Intermediate {
        sender,
        repository,
        organization: _,
        installation,
        mut specific,
    } = serde_json::from_slice::<Intermediate>(body).map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    // Bug: OctoCrab wrongly requires the pusher to have an email
    // Remove when https://github.com/XAMPPRocky/octocrab/issues/486 is fixed
    if kind == WebhookEventType::Push {
        if let Some(pusher) = specific.get_mut("pusher") {
            if let Some(email) = pusher.get_mut("email") {
                if email.is_null() {
                    *email = serde_json::Value::String("".to_owned())
                }
            }
        }
    }

    let specific = kind
        .parse_specific_payload(specific)
        .map_err(|e| (StatusCode::BAD_REQUEST, e.to_string()))?;

    Ok(WebhookEvent {
        sender: sender.map(|s| s.into()),
        repository: repository.map(|r| r.into()),
        installtion_id: match &installation {
            Some(EventInstallation::Full(installation)) => Some(installation.id),
            Some(EventInstallation::Minimal(installation)) => Some(installation.id),
            None => None,
        },
        installation: match installation {
            Some(EventInstallation::Full(installation)) => Some((*installation).into()),
            _ => None,
        },
        kind,
        specific,
    })
}

pub async fn parse_from_request(request: Request<Body>, secret: &str) -> Result<WebhookEvent, (StatusCode, String)> {
    let (parts, body) = request.into_parts();
    let Some(header) = parts.headers.get("X-GitHub-Event").and_then(|v| v.to_str().ok()) else {
        return Err((StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    };

    let Ok(body) = axum::body::to_bytes(body, 1024 * 1024 * 10).await else {
        return Err((StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    };

    if !verify_gh_signature(&parts.headers, &body, secret) {
        return Err((StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    parse_event(header.to_owned(), &body)
}

#[cfg(test)]
#[cfg_attr(coverage_nightly, coverage(off))]
mod tests {
    use super::*;

    #[test]
    fn test_parse_event() {
        let event = parse_event("check_run".to_string(), include_bytes!("mock/webhook.check_run.created.json"))
            .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.created", event);

        let event = parse_event(
            "check_run".to_string(),
            include_bytes!("mock/webhook.check_run.completed.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.check_run.completed", event);

        let event = parse_event("push".to_string(), include_bytes!("mock/webhook.push.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.push", event);

        let event = parse_event(
            "pull_request".to_string(),
            include_bytes!("mock/webhook.pull_request.synchronize.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request.synchronize", event);

        let event = parse_event(
            "issue_comment".to_string(),
            include_bytes!("mock/webhook.issue_comment.created.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.issue_comment.created", event);

        let event = parse_event("delete".to_string(), include_bytes!("mock/webhook.delete.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.delete", event);

        let event = parse_event("create".to_string(), include_bytes!("mock/webhook.create.json")).expect("event parsed");
        insta::assert_debug_snapshot!("webhook.create", event);

        let event = parse_event(
            "pull_request_review".to_string(),
            include_bytes!("mock/webhook.pull_request_review.submitted.json"),
        )
        .expect("event parsed");
        insta::assert_debug_snapshot!("webhook.pull_request_review.submitted", event);
    }

    #[test]
    fn test_verify_gh_signature() {
        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(b"body");
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let mut headers = HeaderMap::new();
        headers.insert("X-Hub-Signature-256", signature.parse().unwrap());
        assert!(verify_gh_signature(&headers, b"body", "secret"));
    }

    #[tokio::test]
    async fn test_parse_from_request() {
        let data = include_bytes!("mock/webhook.push.json");

        let mut mac = Hmac::<Sha256>::new_from_slice("secret".as_bytes()).expect("Cannot create HMAC key");
        mac.update(data);
        let signature = mac.finalize().into_bytes().to_vec();
        let signature = hex::encode(signature);
        let signature = format!("sha256={}", signature);

        let request = Request::builder()
            .header("X-Hub-Signature-256", signature)
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();

        parse_from_request(request, "secret").await.expect("event parsed");
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_signature() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder()
            .header("X-GitHub-Event", "push")
            .body(Body::from(data.to_vec()))
            .unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::UNAUTHORIZED, "Invalid signature".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_missing_header() {
        let data = include_bytes!("mock/webhook.push.json");
        let request = Request::builder().body(Body::from(data.to_vec())).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Missing X-GitHub-Event header".to_string()));
    }

    #[tokio::test]
    async fn test_parse_from_request_invalid_body() {
        let body = Body::from(vec![0; 1024 * 1024 * 20]);
        let request = Request::builder().header("X-GitHub-Event", "push").body(body).unwrap();
        let err = parse_from_request(request, "secret").await.expect_err("error");
        assert_eq!(err, (StatusCode::BAD_REQUEST, "Failed to read body".to_string()));
    }
}
