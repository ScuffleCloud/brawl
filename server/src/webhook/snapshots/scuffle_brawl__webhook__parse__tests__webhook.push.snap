---
source: server/src/webhook/parse.rs
expression: event
snapshot_kind: text
---
WebhookEvent {
    sender: Some(
        User {
            login: "TroyKomodo",
            id: UserId(
                49777269,
            ),
        },
    ),
    repository: Some(
        Repository {
            id: RepositoryId(
                900225914,
            ),
            name: "brawl",
            owner: User {
                login: "ScuffleCloud",
                id: UserId(
                    122814584,
                ),
            },
        },
    ),
    installtion_id: Some(
        InstallationId(
            58286822,
        ),
    ),
    installation: None,
    kind: Push,
    specific: Push(
        PushWebhookEventPayload {
            enterprise: None,
            after: "e48e2992625296b086ffc6430bdefb06894144d9",
            base_ref: None,
            before: "d22a5bdc9af3d51aa6ecc6011b440f0dcda466a3",
            commits: [
                PushWebhookEventCommit {
                    added: [
                        "server/src/github/mock/comment.json",
                        "server/src/github/mock/commit.json",
                        "server/src/github/mock/create_commit.json",
                        "server/src/github/mock/create_merge.json",
                        "server/src/github/mock/create_merge_conflict.json",
                        "server/src/github/mock/delete_branch.json",
                        "server/src/github/mock/delete_branch_non_exist.json",
                        "server/src/github/mock/pr.json",
                        "server/src/github/mock/pr_reviewers.json",
                        "server/src/github/mock/ref_latest_commit.json",
                        "server/src/github/mock/ref_not_found.json",
                        "server/src/github/mock/role_members.json",
                        "server/src/github/mock/team_members.json",
                        "server/src/github/mock/team_not_found.json",
                        "server/src/github/mock/update_ref.json",
                        "server/src/github/mock/user.json",
                        "server/src/github/mock/user_not_found.json",
                        "server/src/utils.rs",
                    ],
                    author: GitUserTime {
                        user: CommitAuthor {
                            name: "Troy Benson",
                            email: "troybensonsa@gmail.com",
                            date: None,
                        },
                        date: None,
                        username: Some(
                            "TroyKomodo",
                        ),
                    },
                    committer: GitUserTime {
                        user: CommitAuthor {
                            name: "Troy Benson",
                            email: "troybensonsa@gmail.com",
                            date: None,
                        },
                        date: None,
                        username: Some(
                            "TroyKomodo",
                        ),
                    },
                    distinct: true,
                    id: "e48e2992625296b086ffc6430bdefb06894144d9",
                    message: "github testing",
                    modified: [
                        "Cargo.lock",
                        "server/Cargo.toml",
                        "server/src/auto_start.rs",
                        "server/src/bin/server.rs",
                        "server/src/command/cancel.rs",
                        "server/src/command/dry_run.rs",
                        "server/src/command/merge.rs",
                        "server/src/command/mod.rs",
                        "server/src/command/ping.rs",
                        "server/src/command/pr.rs",
                        "server/src/command/retry.rs",
                        "server/src/github/config.rs",
                        "server/src/github/installation.rs",
                        "server/src/github/merge_workflow.rs",
                        "server/src/github/messages/mod.rs",
                        "server/src/github/mod.rs",
                        "server/src/github/models.rs",
                        "server/src/github/repo.rs",
                        "server/src/lib.rs",
                        "server/src/webhook/mod.rs",
                    ],
                    removed: [],
                    timestamp: 2024-12-21T23:37:06Z,
                    tree_id: "2736c601a8314842dc50ca679a9ba4b9fa06e374",
                    url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "github.com",
                            ),
                        ),
                        port: None,
                        path: "/ScuffleCloud/brawl/commit/e48e2992625296b086ffc6430bdefb06894144d9",
                        query: None,
                        fragment: None,
                    },
                },
            ],
            compare: Url {
                scheme: "https",
                cannot_be_a_base: false,
                username: "",
                password: None,
                host: Some(
                    Domain(
                        "github.com",
                    ),
                ),
                port: None,
                path: "/ScuffleCloud/brawl/compare/d22a5bdc9af3...e48e29926252",
                query: None,
                fragment: None,
            },
            created: false,
            deleted: false,
            forced: false,
            head_commit: Some(
                PushWebhookEventCommit {
                    added: [
                        "server/src/github/mock/comment.json",
                        "server/src/github/mock/commit.json",
                        "server/src/github/mock/create_commit.json",
                        "server/src/github/mock/create_merge.json",
                        "server/src/github/mock/create_merge_conflict.json",
                        "server/src/github/mock/delete_branch.json",
                        "server/src/github/mock/delete_branch_non_exist.json",
                        "server/src/github/mock/pr.json",
                        "server/src/github/mock/pr_reviewers.json",
                        "server/src/github/mock/ref_latest_commit.json",
                        "server/src/github/mock/ref_not_found.json",
                        "server/src/github/mock/role_members.json",
                        "server/src/github/mock/team_members.json",
                        "server/src/github/mock/team_not_found.json",
                        "server/src/github/mock/update_ref.json",
                        "server/src/github/mock/user.json",
                        "server/src/github/mock/user_not_found.json",
                        "server/src/utils.rs",
                    ],
                    author: GitUserTime {
                        user: CommitAuthor {
                            name: "Troy Benson",
                            email: "troybensonsa@gmail.com",
                            date: None,
                        },
                        date: None,
                        username: Some(
                            "TroyKomodo",
                        ),
                    },
                    committer: GitUserTime {
                        user: CommitAuthor {
                            name: "Troy Benson",
                            email: "troybensonsa@gmail.com",
                            date: None,
                        },
                        date: None,
                        username: Some(
                            "TroyKomodo",
                        ),
                    },
                    distinct: true,
                    id: "e48e2992625296b086ffc6430bdefb06894144d9",
                    message: "github testing",
                    modified: [
                        "Cargo.lock",
                        "server/Cargo.toml",
                        "server/src/auto_start.rs",
                        "server/src/bin/server.rs",
                        "server/src/command/cancel.rs",
                        "server/src/command/dry_run.rs",
                        "server/src/command/merge.rs",
                        "server/src/command/mod.rs",
                        "server/src/command/ping.rs",
                        "server/src/command/pr.rs",
                        "server/src/command/retry.rs",
                        "server/src/github/config.rs",
                        "server/src/github/installation.rs",
                        "server/src/github/merge_workflow.rs",
                        "server/src/github/messages/mod.rs",
                        "server/src/github/mod.rs",
                        "server/src/github/models.rs",
                        "server/src/github/repo.rs",
                        "server/src/lib.rs",
                        "server/src/webhook/mod.rs",
                    ],
                    removed: [],
                    timestamp: 2024-12-21T23:37:06Z,
                    tree_id: "2736c601a8314842dc50ca679a9ba4b9fa06e374",
                    url: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "github.com",
                            ),
                        ),
                        port: None,
                        path: "/ScuffleCloud/brawl/commit/e48e2992625296b086ffc6430bdefb06894144d9",
                        query: None,
                        fragment: None,
                    },
                },
            ),
            pusher: GitUserTime {
                user: CommitAuthor {
                    name: "TroyKomodo",
                    email: "",
                    date: None,
                },
                date: None,
                username: None,
            },
            ref: "refs/heads/troy/refactor",
        },
    ),
}
